File: uikit/src/main/java/com/sendbird/uikit/widgets/PagerRecyclerView.java
package com.sendbird.uikit.widgets;

import android.content.Context;
import android.util.AttributeSet;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

public class PagerRecyclerView extends ThemeableRecyclerView {
    private LinearLayoutManager layoutManager;
    private final OnScrollListener scrollListener = new OnScrollListener();

    public PagerRecyclerView(@NonNull Context context) {
        super(context);
    }

    public PagerRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public PagerRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    @Override
    public void setLayoutManager(LayoutManager layoutManager) {
        if (!(layoutManager instanceof LinearLayoutManager)) throw new IllegalArgumentException("LinearLayoutManager supports only.");
        this.layoutManager = (LinearLayoutManager) layoutManager;
        if (this.scrollListener != null) {
            this.scrollListener.setLayoutManager(this.layoutManager);
        }
        super.setLayoutManager(layoutManager);
    }

    @Override
    public boolean performClick() {
        return super.performClick();
    }

    @Nullable
    @Override
    public LinearLayoutManager getLayoutManager() {
        return layoutManager;
    }

    public void setPager(@NonNull Pageable<?> pager) {
        this.scrollListener.setPager(pager);
        this.scrollListener.setLayoutManager(this.layoutManager);
        addOnScrollListener(this.scrollListener);
    }

    public void setThreshold(int threshold) {
        this.scrollListener.setThreshold(threshold);
    }

    public int findFirstVisibleItemPosition() {
        return layoutManager.findFirstVisibleItemPosition();
    }

    public int findLastVisibleItemPosition() {
        return layoutManager.findLastVisibleItemPosition();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        removeOnScrollListener(this.scrollListener);
        this.scrollListener.dispose();
    }

    public void setOnScrollEndDetectListener(@Nullable OnScrollEndDetectListener scrollEndDetectListener) {
        this.scrollListener.setOnScrollEndDetectListener(scrollEndDetectListener);
    }

    private final static class OnScrollListener extends RecyclerView.OnScrollListener {
        private int threshold = 1;
        private Pageable<?> pager;
        private LinearLayoutManager layoutManager;
        private OnScrollEndDetectListener scrollEndDetectListener;
        private final ExecutorService topLoadingWorker = Executors.newSingleThreadExecutor();
        private final ExecutorService bottomLoadingWorker = Executors.newSingleThreadExecutor();
        private final AtomicBoolean topLoading = new AtomicBoolean(false);
        private final AtomicBoolean bottomLoading = new AtomicBoolean(false);

        public OnScrollListener() {
        }

        public void setPager(@NonNull Pageable<?> pager) {
            this.pager = pager;
        }

        public void setLayoutManager(@Nullable LinearLayoutManager layoutManager) {
            this.layoutManager = layoutManager;
        }

        public void setOnScrollEndDetectListener(@Nullable OnScrollEndDetectListener scrollEndDetectListener) {
            this.scrollEndDetectListener = scrollEndDetectListener;
        }

        @Override
        public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
            int lastVisibleItemPosition = layoutManager.findLastVisibleItemPosition();
            int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();
            int itemCount = layoutManager.getItemCount();

            if (!recyclerView.canScrollVertically(ScrollDirection.Bottom.getDirection())) {
                if (scrollEndDetectListener != null) {
                    scrollEndDetectListener.onScrollEnd(ScrollDirection.Bottom);
                }
            }

            if (!recyclerView.canScrollVertically(ScrollDirection.Top.getDirection())) {
                if (scrollEndDetectListener != null) {
                    scrollEndDetectListener.onScrollEnd(ScrollDirection.Top);
                }
            }

            if (pager == null) return;
            final boolean reverseLayout = layoutManager.getReverseLayout();
            final boolean topLoadMore = pager.hasPrevious() && (reverseLayout ? itemCount - lastVisibleItemPosition <= threshold : firstVisibleItemPosition <= threshold);
            if (!topLoading.get() && topLoadMore) {
                topLoading.set(true);
                topLoadingWorker.submit(() -> {
                    try {
                        pager.loadPrevious();
                    } catch (Exception ignore) {
                    } finally {
                        topLoading.set(false);
                    }
                });
            }
            //final boolean bottomLoadMore = !(reverseLayout? !pager.hasNext() : !pager.hasPrevious());
            final boolean bottomLoadMore = pager.hasNext() && (reverseLayout ? firstVisibleItemPosition <= threshold : itemCount - lastVisibleItemPosition <= threshold);
            if (!bottomLoading.get() && bottomLoadMore) {
                bottomLoading.set(true);
                bottomLoadingWorker.submit(() -> {
                    try {
                        pager.loadNext();
                    } catch (Exception ignore) {
                    } finally {
                        bottomLoading.set(false);
                    }
                });
            }
        }

        public void setThreshold(int threshold) {
            if (threshold <= 0) {
                throw new IllegalArgumentException("illegal threshold: " + threshold);
            }
            this.threshold = threshold;
        }

        public void dispose() {
            this.topLoadingWorker.shutdown();
            this.bottomLoadingWorker.shutdown();
        }
    }

    public interface Pageable<T> {
        /**
         * Synchronized function call must be used.
         */
        T loadPrevious() throws Exception;

        /**
         * Synchronized function call must be used.
         */
        T loadNext() throws Exception;

        boolean hasNext();
        boolean hasPrevious();
    }

    public interface OnScrollEndDetectListener {
        void onScrollEnd(ScrollDirection direction);
    }

    public enum ScrollDirection {
        Top(-1), Bottom(1);
        private final int value;
        ScrollDirection(int value) {
            this.value = value;
        }

        public int getDirection() {
            return value;
        }
    }
}

File: uikit/src/main/java/com/sendbird/uikit/internal/ui/widgets/PagerRecyclerView.kt
package com.sendbird.uikit.internal.ui.widgets

import android.content.Context
import android.util.AttributeSet
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.sendbird.uikit.interfaces.OnPagedDataLoader
import org.jetbrains.annotations.TestOnly
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

internal class InnerLinearLayoutManager(context: Context) : LinearLayoutManager(context) {
    var listener: OnLayoutCompleteListener? = null

    override fun onLayoutCompleted(state: RecyclerView.State) {
        super.onLayoutCompleted(state)
        listener?.onLayoutComplete(state)
    }
}

internal interface OnLayoutCompleteListener {
    fun onLayoutComplete(state: RecyclerView.State)
}

internal class PagerRecyclerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : ThemeableRecyclerView(context, attrs, defStyle) {
    private var layoutManager: LinearLayoutManager? = null
    private val onScrollListener: OnScrollListener by lazy { OnScrollListener(layoutManager) }
    var onLayoutCompleteListener: OnLayoutCompleteListener? = null
        set(value) {
            field = value
            if (layoutManager is InnerLinearLayoutManager) (layoutManager as InnerLinearLayoutManager).listener = value
        }

    override fun setLayoutManager(layoutManager: LayoutManager?) {
        require(layoutManager is LinearLayoutManager) { "LinearLayoutManager supports only." }
        this.layoutManager = layoutManager
        super.setLayoutManager(layoutManager)
    }

    override fun getLayoutManager(): LinearLayoutManager? = layoutManager

    fun setPager(pager: OnPagedDataLoader<*>) {
        onScrollListener.pager = pager
        onScrollListener.layoutManager = layoutManager
        addOnScrollListener(onScrollListener)
    }

    @TestOnly
    fun getPager() = onScrollListener.pager

    fun setThreshold(threshold: Int) {
        onScrollListener.threshold = threshold
    }

    @TestOnly
    fun getThreshold(): Int = onScrollListener.threshold

    fun findFirstVisibleItemPosition(): Int {
        return layoutManager?.findFirstVisibleItemPosition() ?: 0
    }

    fun findLastVisibleItemPosition(): Int {
        return layoutManager?.findLastVisibleItemPosition() ?: 0
    }

    fun isReverseLayout(): Boolean {
        return layoutManager?.reverseLayout ?: false
    }

    fun setOnScrollEndDetectListener(scrollEndDetectListener: OnScrollEndDetectListener?) {
        onScrollListener.scrollEndDetectListener = scrollEndDetectListener
    }

    fun useReverseData() {
        onScrollListener.useReverseData = true
    }

    fun setStackFromEnd(stackFromEnd: Boolean) {
        post { layoutManager?.stackFromEnd = stackFromEnd }
    }

    fun isScrollable(): Boolean {
        return canScrollVertically(1) || canScrollVertically(-1)
    }

    fun getStackFromEnd(): Boolean = layoutManager?.stackFromEnd ?: false

    override fun performClick(): Boolean {
        return super.performClick()
    }

    private class OnScrollListener constructor(var layoutManager: LinearLayoutManager?) :
        RecyclerView.OnScrollListener() {
        var threshold = 1
            set(value) {
                require(value > 0) { "illegal threshold: $value" }
                field = value
            }
        var pager: OnPagedDataLoader<*>? = null
        var scrollEndDetectListener: OnScrollEndDetectListener? = null
        private val prevLoadingWorker = Executors.newSingleThreadExecutor()
        private val nextLoadingWorker = Executors.newSingleThreadExecutor()
        private val prevLoading = AtomicBoolean(false)
        private val nextLoading = AtomicBoolean(false)
        var useReverseData: Boolean = false

        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
            layoutManager?.run {
                val lastVisibleItemPosition = findLastVisibleItemPosition()
                val firstVisibleItemPosition = findFirstVisibleItemPosition()
                if (!recyclerView.canScrollVertically(ScrollDirection.Bottom.direction)) {
                    scrollEndDetectListener?.onScrollEnd(ScrollDirection.Bottom)
                }
                if (!recyclerView.canScrollVertically(ScrollDirection.Top.direction)) {
                    scrollEndDetectListener?.onScrollEnd(ScrollDirection.Top)
                }
                pager?.run {
                    val prevLoadMore =
                        hasPrevious() && if (useReverseData) itemCount - lastVisibleItemPosition <= threshold else firstVisibleItemPosition <= threshold
                    if (!prevLoading.get() && prevLoadMore) {
                        prevLoading.set(true)
                        prevLoadingWorker.submit {
                            try {
                                loadPrevious()
                            } catch (ignore: Exception) {
                            } finally {
                                prevLoading.set(false)
                            }
                        }
                    }

                    val nextLoadMore =
                        hasNext() && if (useReverseData) firstVisibleItemPosition <= threshold else itemCount - lastVisibleItemPosition <= threshold
                    if (!nextLoading.get() && nextLoadMore) {
                        nextLoading.set(true)
                        nextLoadingWorker.submit {
                            try {
                                loadNext()
                            } catch (ignore: Exception) {
                            } finally {
                                nextLoading.set(false)
                            }
                        }
                    }
                }
            }
        }

        fun dispose() {
            prevLoadingWorker.shutdown()
            nextLoadingWorker.shutdown()
        }
    }

    interface OnScrollEndDetectListener {
        fun onScrollEnd(direction: ScrollDirection)
    }

    enum class ScrollDirection(val direction: Int) {
        Top(-1), Bottom(1);
    }
}
